### ISSUE TITLE
How do I count the number of datapoints in a plot?

### ISSUE THREAD

**Issue by fonsp** [JuliaPlots/Plots.jl#3437]
I am adding specialized support for GR-backend-Plots.jl into Pluto.jl ([PR](https://github.com/fonsp/Pluto.jl/pull/1090)). One optimization is that I want to use SVG output for most GR-backed plots, but PNG if the number of datapoints is to large, to avoid overloading the browser.

I currently implemented it like this:
```julia
approx_size(p::Plots.Plot) = try
    sum(p.series_list) do series
        length(series[:y])
    end
catch e
    @warn "Failed to estimate plot size" exception=(e,catch_backtrace())
    0
end
```

Is this a good approach? Is there a Plots.jl feature that I could use for this?

_Thanks in advance, I enjoy Plots.jl, I want to make it easier for Pluto users to plot their data, and I hope we can work on further integrations in the future! -fons_

**Comment by isentropic** [comment]
It always bothered me why GR even defaults to svg format instead of png. Do you think that

**Comment by fonsp** [comment]
@isentropic Can you check your last message?

**Comment by daschw** [comment]
Hi @fonsp ! I like the idea and could not think of another way than counting the data points in series_list. We don't have that feature yet, but maybe we could add it. We could even think of defaulting to `plot(..., fmt=:auto)` and choose PNG above a certain threshold of `approx_size` in Plots. I have not really thought that through, though. So, it's probably best to stay with your Pluto-specific implementation in https://github.com/fonsp/Pluto.jl/pull/1090.

**Comment by isentropic** [comment]
Right, pardon for the unfinished comment, I accidentally pressed "Close with comment" without even finishing my point.

It always bothered me why GR even defaults to svg format instead of png. I agree that png is only favorable when the svg becomes too heavy (both size and render speed). I like what @daschw has suggested with`fmt:auto`

Instead of counting the points in the series (perhaps there are a lot of subplots with multiple series), I suggest that we check for the implemented svg size in megabytes as a callback. If the rendered image svg exceeds some size, it could instead be instructed to render png. This I believe could be done in a backend agnostic way:
1. Render svg by default, check the generated svg size in MB
2. Rerender in png in case the svg is too heavy. (Perhaps throw an `@info` informing the user that the picture is no longer svg due to size considerations.

**Comment by isentropic** [comment]
I'm not sure how Plots could easily access the notebook (Pluto or Jupyter) svg file size. Perhaps it could be GR specific for the time being, or `fmt:auto` should mean that the backend should decide appropriately.

**Comment by fonsp** [comment]
@isentropic I also considered this approach (i.e. rendering SVG and checking size), but I could not figure out how to make that work without rendering the plot twice, once to determine whether SVG should be used, and once to actually generate the plot (be that in SVG or PNG).

Of course, it is _possible_, but I would have to break out of the current display system, which is:
1. use `Base.showable` to go through the list of supported mimes.
2. use `Base.show` to render the object.

**Comment by isentropic** [comment]
I think that counting the number of points in the series is not all that robust. Let's say its a simple line `X=Y` plot, with let's say 1e6 points. I'm sure the rendering backend is smart enough to reduce the over-specification (downsampling) of the line plot: the plotting backend does not render all the points, only the points that are meaningful to show. So the 1e6 point line plot would end up being much compact after a backend renders it into a file. Plots series data is perhaps the most inefficient storage without any compression or downsampling: oftentimes it would not represent the real size of the svg (or pdf, png).

**Comment by isentropic** [comment]
Or, if there are multiple markers that practically cover each other completely, backends (at least matplotlib does it) would perform simplifications and downsampling.
Perhaps this would illustrate what I am talking about:
https://stackoverflow.com/questions/7454159/gnuplot-x-axis-resolution

**Comment by isentropic** [comment]
Judging in terms of the amount in the series of each subplot, would not be all that robust. If simple enough shapes are plotted (or there is too many finely spaced points) smart backends (I assume all are, as even Gnuplot got it) would undersample when necessary and drastically reduce the rendered file size

**Comment by isentropic** [comment]
Another way would be to always raster the rendered svg plot (figures) but not the texts --- https://www.astrobetter.com/blog/2014/01/17/slim-down-your-bloated-graphics/

If there is a simple way to do this in GR, then I think it would be a good default.
